### ISSUE TITLE
native-image crashes unpredictably

### ISSUE THREAD

**Issue by mthmulders** [oracle/graal#854]
I was trying to build a native executable for my Java program. The program happens to be a launcher similar to that of the Simple Language repository: it runs a guest language program. In my case, the guest language is Brainfuck. Complete source code is [here](https://github.com/mthmulders/brainfuck-jvm).

When I build the project on my Mac, everything runs fine. When I run it on a CI environment, CircleCI in my case, generating the native image fails. Manually retrying makes it succeed (around 1 out of 6 attemts) but the regular CI builds keep on failing. CircleCI runs builds in Linux containers.

I compared verbose logging output of both `native-tool` invocations. There aren't many relevant differences. One of them is the fact that on CircleCI/Linux, the `-watchpid <some-number>` argument is added. Looking at the source code for native-image, I found that it depends on `-Dcom.oracle.graalvm.isaot=true` and on the fact that the OS uses procfs. Since I can't change the latter, [I changed the former](https://github.com/mthmulders/brainfuck-jvm/commit/09170d5f1571982b79631383fb9f90619e239201) but I'm not sure if this is the way to go.

In summary: I expect random, unpredictable crashes aren't the expected behavior of `native-image`. If this is indeed a bug, I'm happy to share any more details that might be needed to solve it.

**Comment by mthmulders** [comment]
Failed CI builds:
* https://circleci.com/gh/mthmulders/brainfuck-jvm/24 (with `-Dcom.oracle.graalvm.isaot=true)
* https://circleci.com/gh/mthmulders/brainfuck-jvm/25 (with `-Dcom.oracle.graalvm.isaot=false)

So it seems that even `-Dcom.oracle.graalvm.isaot` doesn't make a difference here...

**Comment by cstancu** [comment]
@mthmulders I see from the logs that you are using rc9. Can you please try using rc10 and see if the problem is persistent?

**Comment by mthmulders** [comment]
I wish, but rc10 is still not available in [Maven Central](https://search.maven.org/search?q=graal-sdk).

**Comment by mthmulders** [comment]
Today, I've upgraded the project to RC10, as the artifacts have arrived in Maven Central. No luck, the problem still persists.

**Comment by mthmulders** [comment]
[Still happening](https://circleci.com/gh/mthmulders/brainfuck-jvm/50) with RC11...

**Comment by cstancu** [comment]
@mthmulders the stack trace in the logs, i.e.,  `org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)`, is not from the `native-image` tool, but rather from your CI. Can you re-run Maven with the -e switch to get to see a full stack trace? I don't know how many resources you allocate in your CI (memory, CPU time), but it may be that `native-image` exhausts the allocated resources and the CI terminates it.

**Comment by mthmulders** [comment]
@cstancu I upgraded to RC12 in the meantime. The problem is still there. Full log is [here](https://circleci.com/gh/mthmulders/brainfuck-jvm/57).

I think the relevant part is this:

    Executing [
    /opt/graalvm-ce-1.0.0-rc12/bin/java \
    -XX:+UnlockExperimentalVMOptions \
    -XX:+EnableJVMCI \
    -XX:-UseJVMCICompiler \
    -Dtruffle.TrustAllTruffleRuntimeProviders=true \
    -d64 \
    -XX:-UseJVMCIClassLoader \
    -Xss10m \
    -Xms1g \
    -Xmx14g \
    -Duser.country=US \
    -Duser.language=en \
    -Dgraalvm.version=1.0.0-rc12 \
    -Dorg.graalvm.version=1.0.0-rc12 \
    -Dcom.oracle.graalvm.isaot=true \
    -Djvmci.class.path.append=/opt/graalvm-ce-1.0.0-rc12/jre/lib/jvmci/graal.jar \
    -Dgraalvm.locatorDisabled=true \
    -Dtruffle.TrustAllTruffleRuntimeProviders=true \
    -Xbootclasspath/a:/opt/graalvm-ce-1.0.0-rc12/jre/lib/boot/graaljs-scriptengine.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/boot/graal-sdk.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/truffle/truffle-api.jar \
    -cp \
    /opt/graalvm-ce-1.0.0-rc12/jre/lib/svm/builder/pointsto.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/svm/builder/svm.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/svm/builder/objectfile.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/jvmci/graal-management.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/jvmci/jvmci-api.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/jvmci/jvmci-hotspot.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/jvmci/graal.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/truffle/truffle-nfi.jar \
    com.oracle.svm.hosted.NativeImageGeneratorRunner \
    -watchpid \
    566 \
    -imagecp \
    /opt/graalvm-ce-1.0.0-rc12/jre/lib/boot/graaljs-scriptengine.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/boot/graal-sdk.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/truffle/truffle-api.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/svm/builder/pointsto.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/svm/builder/svm.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/svm/builder/objectfile.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/jvmci/graal-management.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/jvmci/jvmci-api.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/jvmci/jvmci-hotspot.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/jvmci/graal.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/truffle/truffle-nfi.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/svm/library-support.jar:/opt/graalvm-ce-1.0.0-rc12/jre/lib/graalvm/*:/root/project/language/target/brainfuck-jvm.jar:/root/project/launcher/target/launcher.jar \
    -H:Path=/root/project/native \
    -H:Features=com.oracle.svm.truffle.TruffleFeature,com.oracle.svm.truffle.nfi.TruffleNFIFeature \
    -H:+ReportUnsupportedElementsAtRuntime \
    -H:MaxRuntimeCompileMethods=1900 \
    -H:CLibraryPath=/opt/graalvm-ce-1.0.0-rc12/jre/lib/svm/clibraries/linux-amd64,/opt/graalvm-ce-1.0.0-rc12/jre/lib/truffle \
    -H:Class=it.mulders.brainfuckjvm.launcher.Launcher \
    -H:Name=target/bfnative
    ]
    [target/bfnative:567]    classlist:   1,127.46 ms
    [target/bfnative:567]        (cap):     598.30 ms
    [target/bfnative:567]        setup:   1,456.48 ms
    [target/bfnative:567]   (typeflow):  11,257.55 ms
    [target/bfnative:567]    (objects):  18,212.12 ms
    [target/bfnative:567]   (features):     697.79 ms
    [target/bfnative:567]     analysis:  30,866.32 ms
    805 method(s) included for runtime compilation
    [target/bfnative:567]     universe:     759.54 ms
    [target/bfnative:567]      (parse):   2,479.07 ms
    [target/bfnative:567]     (inline):   1,744.10 ms
    Error: Image building with exit status 137
    [ERROR] Command execution failed.
    org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)

**Comment by cstancu** [comment]
The `Error: Image building with exit status 137` means that the image building process was forcefully stopped, it is essentially equivalent with `kill -9 (128 + 9)`. I think what happens is that the [process runs out of the available memory and the CI kills it](https://support.circleci.com/hc/en-us/articles/115014359648-Exit-code-137-Out-of-memory).

**Comment by mthmulders** [comment]
OK, seems like a limitation in my CI plan. Thanks for pointing this out!
